File: C++/023_MergeKSortedLinkedList_PriorityQueue.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
/*
Computation: O(nlog(k))
Space: O(k)
*/
struct cmp{
    bool operator()(ListNode*a, ListNode*b){
        return a->val < b->val;
    }  
};

class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        priority_queue<ListNode*, vector<ListNode*>, cmp> q;
        
        for(int i=0; i< lists.size(); i++){
            if(lists[i]) q.push(lists[i]);
        }
        // qtop is the pointer to the top of the prioity queue
        // res is the pointer to be returned
        // tmp is the pointer helping to build the link
        ListNode *qtop=NULL, *res = NULL, *tmp=NULL; 
        while( !q.empty()){
            // get the top node
            qtop = q.top(); q.pop();
            
            // update link
            if(!tmp) 
                res = qtop;
            else 
                tmp->next =qtop;
            tmp = qtop;
            
            // update the queue
            if(qtop->next)
                q.push(qtop->next);
        }
        return res;
    }
};

File: Python/023_MergeKSortedLinkedList_PriorityQueue.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

import heapq
class Solution(object):
    def mergeKLists(self, lists):
        """
        :type lists: List[ListNode]
        :rtype: ListNode
        """
        q = []
        for l in lists:
            if l:
                heapq.heappush(q, (l.val, l))
        
        res = ListNode(0)
        tmp = res
        
        while q:
            qtop = heapq.heappop(q)[1]
            tmp.next = qtop
            tmp = tmp.next
            if qtop.next:
                heapq.heappush(q, (qtop.next.val, qtop.next))
        
        return res.next